WordPress/src/test/java/org/wordpress/android/ui/jetpack/backup/download/usecases/GetBackupDownloadStatusUseCaseTest.kt

```diff
@@ -0,0 +1,128 @@
+package org.wordpress.android.ui.jetpack.backup.download.usecases
+
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.flow.toList
+import org.assertj.core.api.Assertions
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mock
+import org.wordpress.android.BaseUnitTest
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.model.activity.BackupDownloadStatusModel
+import org.wordpress.android.fluxc.store.ActivityLogStore
+import org.wordpress.android.fluxc.store.ActivityLogStore.OnBackupDownloadStatusFetched
+import org.wordpress.android.test
+import org.wordpress.android.ui.jetpack.backup.download.BackupDownloadRequestState.Failure
+import org.wordpress.android.util.NetworkUtilsWrapper
+import org.wordpress.android.fluxc.action.ActivityLogAction.FETCH_BACKUP_DOWNLOAD_STATE
+import org.wordpress.android.fluxc.store.ActivityLogStore.BackupDownloadStatusError
+import org.wordpress.android.fluxc.store.ActivityLogStore.BackupDownloadStatusErrorType.GENERIC_ERROR
+import org.wordpress.android.ui.jetpack.backup.download.BackupDownloadRequestState.Complete
+import org.wordpress.android.ui.jetpack.backup.download.BackupDownloadRequestState.Empty
+import org.wordpress.android.ui.jetpack.backup.download.BackupDownloadRequestState.Failure.RemoteRequestFailure
+import org.wordpress.android.ui.jetpack.backup.download.BackupDownloadRequestState.Progress
+import java.util.Date
+
+@InternalCoroutinesApi
+class GetBackupDownloadStatusUseCaseTest : BaseUnitTest() {
+    private lateinit var useCase: GetBackupDownloadStatusUseCase
+    @Mock lateinit var networkUtilsWrapper: NetworkUtilsWrapper
+    @Mock lateinit var activityLogStore: ActivityLogStore
+    @Mock private lateinit var site: SiteModel
+    @Mock lateinit var statusModel: BackupDownloadStatusModel
+
+    @Before
+    fun setup() {
+        useCase = GetBackupDownloadStatusUseCase(networkUtilsWrapper, activityLogStore)
+        whenever(networkUtilsWrapper.isNetworkAvailable()).thenReturn(true)
+        whenever(activityLogStore.getBackupDownloadStatusForSite(site)).thenReturn(statusModel)
+    }
+
+    @Test
+    fun `given site, when status is fetched, then NetworkUnavailable is returned on no network`() = test {
+        whenever(networkUtilsWrapper.isNetworkAvailable()).thenReturn(false)
+
+        val result = useCase.getBackupDownloadStatus(site, downloadId).toList(mutableListOf())
+
+        Assertions.assertThat(result).contains(Failure.NetworkUnavailable)
+    }
+
+    @Test
+    fun `given site, when status is fetched, then RemoteRequestFailure is returned on failure`() = test {
```

Comment:
Warning (⚠️): This test will always run with a delay of 5 seconds, given the fact that the `DELAY_MILLIS` is set to this value. This means that this test suite will deteriorate how long the tests are run in general.

Consider using the `MainCoroutineScopeRule` approach to fake the delay, by replacing `test { ... }` with `coroutineScope.runBlockingTest { ... }` and then add the `advanceTimeBy(DELAY_MILLIS)` right after the test `act` line. Let me know if you need help with that and I can explain in more detail.

PS: The same applies for the `progress` related test at the end which also delays to completion.

Reply:
Nice catch. 🙇

Reply:
👋 @zwarm !

Thanks for making the change. 🙏

May I suggest going with using the `coroutineScope.runBlockingTest { ... }` instead of using `runBlockingTest { ... }` directly, as per my comment above?

I only insist on that for two reasons:
1) Using the `coroutineScope` to trigger the `runBlockingTest { ... }` helps making sure that all `runBlockingTest { ... }` related tests are running as part of the test scope and so they don't compete with each other, which makes the test more coroutine robust.
2) Searching for `runBlockingTest { ... }` on the whole `WordPress-Android` repo, I saw that the team has defined this `MainCoroutineScopeRule` and I am guessing this is what we should be using in such case. You can search for it and see how they set up a test with that (there are a couple of test with that pattern and `runBlockingTest` in general so far). @malinajirka maybe you can help us with the history here?

Apologies if I wasn't too clear on my initial comment. Let me know when you come back online and we will deal with that together so that to merge this with develop asap.

Reply:
@ParaskP7 I think I got it. See commit a707713. 😉

Reply:
Awesome, thanks @zwarm , it looks good! 🌟